arangod/IResearch/AqlHelper.h

```diff
@@ -421,11 +421,30 @@ bool visitAttributeAccess(aql::AstNode const*& head, aql::AstNode const* node,
         auto* root = itr->getMemberUnchecked(1);
         auto* var = itr->getMemberUnchecked(0);
 
-        return ref && aql::NODE_TYPE_ITERATOR == itr->type &&
-               aql::NODE_TYPE_REFERENCE == ref->type && var &&
-               aql::NODE_TYPE_VARIABLE == var->type &&
-               visitAttributeAccess(head, root, visitor)  // 1st visit root
-               && visitor.expansion(*node);  // 2nd visit current node
+        if (!ref) {
+          return false;  // malformed node
+        }
+
+        if (aql::NODE_TYPE_REFERENCE == ref->type) {
+          // b.a[*]
+          return aql::NODE_TYPE_ITERATOR == itr->type &&
+                 aql::NODE_TYPE_REFERENCE == ref->type && var &&
+                 aql::NODE_TYPE_VARIABLE == var->type &&
+                 visitAttributeAccess(head, root, visitor)  // 1st visit root
+                 && visitor.expansion(*node);  // 2nd visit current node
+        } else if (aql::NODE_TYPE_ATTRIBUTE_ACCESS == ref->type) {
+          // tail after the expansion b.a[*].c
+          aql::AstNode const* tailHead{nullptr};
+          return aql::NODE_TYPE_ITERATOR == itr->type && var &&
+                 aql::NODE_TYPE_VARIABLE == var->type &&
+                 visitAttributeAccess(head, root, visitor) &&  // 1st visit root
+                 visitor.expansion(*node) &&  // 2nd visit current node
+                 visitAttributeAccess(tailHead, ref,
+                                      visitor);  // 3rd visit the tail
+        } else {                                 // unexpected expansion form
+          TRI_ASSERT(false);
+          return false;
+        }
```

Comment:
If  NODE_TYPE_EXPANSION node structure will be changed and we will forget to update this function we definitely want to have an assertion triggered. And for prod we at least will fail parsing (this will most likely fail the query but in graceful manner)